frame/examples/default-config/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Default Config Pallet Example
+//!
+//! A simple example of a FRAME pallet that utilizes [`frame_support::derive_impl`] to demonstrate
+//! the simpler way to implement `Config` trait of pallets. This example only showcases this in a
+//! `mock.rs` environment, but the same applies to a real runtime as well.
+//!
+//! See the source code of [`tests`] for a real examples.
+//!
+//! Study the following types:
+//!
+//! - [`pallet::DefaultConfig`], and how it differs from [`pallet::Config`].
+//! - [`pallet::config_preludes::TestDefaultConfig`] and how it implements
+//!   [`pallet::DefaultConfig`].
+//! - Notice how [`pallet::DefaultConfig`] is independent of [`frame_system::Config`].
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+
+	/// This pallet is annotated to have a default config. This will auto-generate
+	/// [`DefaultConfig`].
+	///
+	/// It will be an identical, but won't have anything that is `#[pallet::no_default]`.
+	#[pallet::config(with_default)]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type. This is coming from the runtime, and cannot have a default.
+		/// In general, `Runtime*`-oriented types cannot have a sensible default.
+		#[pallet::no_default] // optional. `RuntimeEvent` is automatically excluded as well.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
```

Comment:
Do we do a keyword match here on the identifier to see whether it starts with `Runtime*`? Because that sounds like a bad idea -- we didn't explicitly say that this is a rule that we followed before, and if the user has a config item that just so happens to be called `Runtime*` and they wanted to create a default for it, they'll have no means to do so.

Reply:
It's just referring to `RuntimeEvent` having `#[no_default]` attached to it. I am fairly certain we don't do any keyword/prefix-based assumptions

Reply:
Yeah, `RuntimeEvent` is already matched against and expected to exist EXACTLY in this form in the pallet has `#[pallet::event]`, and can never have a default.

Reply:
in other words, no pattern matching shenanigans is being done here.